File: app/src/main/java/com/github/shadowsocks/AppManager.java
package com.github.shadowsocks;

import android.Manifest;
import android.app.TaskStackBuilder;
import android.content.BroadcastReceiver;
import android.content.ClipData;
import android.content.ClipboardManager;
import android.content.Context;
import android.content.Intent;
import android.content.IntentFilter;
import android.content.pm.PackageInfo;
import android.content.pm.PackageManager;
import android.graphics.drawable.Drawable;
import android.os.Bundle;
import android.os.Handler;
import android.text.TextUtils;
import android.view.KeyEvent;
import android.view.LayoutInflater;
import android.view.MenuItem;
import android.view.View;
import android.view.ViewGroup;
import android.widget.CompoundButton;
import android.widget.ImageView;
import android.widget.Switch;
import android.widget.TextView;

import com.github.shadowsocks.R;
import com.github.shadowsocks.database.Profile;
import com.github.shadowsocks.utils.Constants;
import com.github.shadowsocks.utils.ToastUtils;
import com.github.shadowsocks.utils.Utils;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.Collections;
import java.util.Comparator;
import java.util.HashSet;
import java.util.List;
import java.util.concurrent.atomic.AtomicBoolean;

import androidx.annotation.Nullable;
import androidx.appcompat.app.AppCompatActivity;
import androidx.appcompat.widget.Toolbar;
import androidx.recyclerview.widget.DefaultItemAnimator;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;

public class AppManager extends AppCompatActivity implements Toolbar.OnMenuItemClickListener {

    private static AppManager instance;

    private boolean receiverRegistered;
    private List<ProxiedApp> cachedApps;

    private HashSet<String> proxiedApps;
    private Toolbar toolbar;
    private Switch bypassSwitch;
    private RecyclerView appListView;
    private View loadingView;
    private AtomicBoolean appsLoading;
    private Handler handler;
    private Profile profile = ShadowsocksApplication.app.currentProfile();

    private void initProxiedApps() {
        initProxiedApps(profile.individual);
    }

    private void initProxiedApps(String str) {
        String[] split = str.split("\n");
        List<String> list = Arrays.asList(split);
        proxiedApps = new HashSet<>(list);
    }

    @Override
    protected void onDestroy() {
        instance = null;
        super.onDestroy();

        if (handler != null) {
            handler.removeCallbacksAndMessages(null);
            handler = null;
        }
    }

    @Override
    public boolean onMenuItemClick(MenuItem item) {
        ClipboardManager clipboard = (ClipboardManager) getSystemService(Context.CLIPBOARD_SERVICE);
        String proxiedAppString;
        switch (item.getItemId()) {
            case R.id.action_apply_all:
                List<Profile> profiles = ShadowsocksApplication.app.profileManager.getAllProfiles();
                if (profiles != null) {
                    proxiedAppString = profile.individual;
                    boolean in_proxyapp = profile.proxyApps;
                    boolean in_bypass = profile.bypass;
                    for (Profile p : profiles) {
                        p.individual = proxiedAppString;
                        p.bypass = in_bypass;
                        p.proxyApps = in_proxyapp;
                        ShadowsocksApplication.app.profileManager.updateProfile(p);
                    }
                    ToastUtils.showShort(R.string.action_apply_all);
                } else {
                    ToastUtils.showShort(R.string.action_export_err);
                }
                return true;
            case R.id.action_export:
                boolean bypass = profile.bypass;
                proxiedAppString = profile.individual;
                ClipData clip = ClipData.newPlainText(Constants.Key.individual, bypass + "\n" + proxiedAppString);
                if (clipboard != null) {
                    clipboard.setPrimaryClip(clip);
                    ToastUtils.showShort(R.string.action_export_msg);
                }
                return true;
            case R.id.action_import:
                if (clipboard != null && clipboard.hasPrimaryClip()) {
                    CharSequence proxiedAppSequence = clipboard.getPrimaryClip().getItemAt(0).getText();
                    if (proxiedAppSequence != null) {
                        proxiedAppString = proxiedAppSequence.toString();
                        if (!TextUtils.isEmpty(proxiedAppString)) {
                            int i = proxiedAppString.indexOf('\n');
                            try {
                                String enabled;
                                String apps;
                                if (i < 0) {
                                    enabled = proxiedAppString;
                                    apps = "";
                                } else {
                                    enabled = proxiedAppString.substring(0, i);
                                    apps = proxiedAppString.substring(i + 1);
                                }

                                bypassSwitch.setChecked(Boolean.parseBoolean(enabled));
                                profile.individual = apps;
                                ShadowsocksApplication.app.profileManager.updateProfile(profile);
                                ToastUtils.showShort(R.string.action_import_msg);
                                appListView.setVisibility(View.GONE);
                                loadingView.setVisibility(View.VISIBLE);
                                initProxiedApps(apps);
                                reloadApps();
                                return true;
                            } catch (Exception e) {
                                ToastUtils.showShort(R.string.action_import_err);
                            }
                        }
                    }
                }
                ToastUtils.showShort(R.string.action_import_err);
                return false;
            default:
                break;
        }
        return false;
    }

    @Override
    protected void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        if (profile == null) {
            finish();
        }

        handler = new Handler();
        appsLoading = new AtomicBoolean();

        setContentView(R.layout.layout_apps);
        toolbar = findViewById(R.id.toolbar);
        toolbar.setTitle(R.string.proxied_apps);
        toolbar.setNavigationIcon(R.drawable.abc_ic_ab_back_material);
        toolbar.setNavigationOnClickListener(v -> {
            Intent intent = getParentActivityIntent();
            if (shouldUpRecreateTask(intent) || isTaskRoot()) {
                TaskStackBuilder.create(AppManager.this)
                        .addNextIntentWithParentStack(intent)
                        .startActivities();
            } else {
                finish();
            }
        });
        toolbar.inflateMenu(R.menu.app_manager_menu);
        toolbar.setOnMenuItemClickListener(this);

        if (!profile.proxyApps) {
            profile.proxyApps = true;
            ShadowsocksApplication.app.profileManager.updateProfile(profile);
        }

        ((Switch) findViewById(R.id.onSwitch)).setOnCheckedChangeListener(new CompoundButton.OnCheckedChangeListener() {
            @Override
            public void onCheckedChanged(CompoundButton buttonView, boolean checked) {
                profile.proxyApps = checked;
                ShadowsocksApplication.app.profileManager.updateProfile(profile);
                finish();
            }
        });

        bypassSwitch = findViewById(R.id.bypassSwitch);
        bypassSwitch.setChecked(profile.bypass);
        bypassSwitch.setOnCheckedChangeListener((buttonView, isChecked) -> {
            profile.bypass = isChecked;
            ShadowsocksApplication.app.profileManager.updateProfile(profile);
        });
        initProxiedApps();
        loadingView = findViewById(R.id.loading);
        appListView = findViewById(R.id.applistview);
        appListView.setLayoutManager(new LinearLayoutManager(this));
        appListView.setItemAnimator(new DefaultItemAnimator());

        instance = this;

        new Thread(() -> loadAppsAsync()).start();

    }


    private void reloadApps() {
        if (!appsLoading.compareAndSet(true, false)) {
            loadAppsAsync();
        }
    }

    private void loadAppsAsync() {
        if (!appsLoading.compareAndSet(false, true)) {
            return;
        }
        AppsAdapter tempAdapter;
        do {
            appsLoading.set(true);
            tempAdapter = new AppsAdapter();
        } while (!appsLoading.compareAndSet(true, false));

        final AppsAdapter adapter = tempAdapter;
        handler.post(() -> {
            appListView.setAdapter(adapter);
            Utils.crossFade(AppManager.this, loadingView, appListView);
        });
    }

    @Override
    public boolean onKeyUp(int keyCode, KeyEvent event) {
        if (keyCode == KeyEvent.KEYCODE_MENU) {
            if (toolbar.isOverflowMenuShowing()) {
                return toolbar.hideOverflowMenu();
            } else {
                return toolbar.showOverflowMenu();
            }
        }
        return super.onKeyUp(keyCode, event);
    }

    private List<ProxiedApp> getApps(PackageManager pm) {
        if (!receiverRegistered) {
            IntentFilter filter = new IntentFilter(Intent.ACTION_PACKAGE_ADDED);
            filter.addAction(Intent.ACTION_PACKAGE_REMOVED);
            filter.addDataScheme("package");
            ShadowsocksApplication.app.registerReceiver(new BroadcastReceiver() {
                @Override
                public void onReceive(Context context, Intent intent) {
                    if (Intent.ACTION_PACKAGE_REMOVED.equals(intent.getAction()) || !intent.getBooleanExtra(Intent.EXTRA_REPLACING, false)) {
                        synchronized (ProxiedApp.class) {
                            AppManager instance = AppManager.instance;
                            if (instance != null) {
                                instance.reloadApps();
                            }
                        }
                    }
                }
            }, filter);
            receiverRegistered = true;
        }

        synchronized (AppManager.class) {
            if (cachedApps == null) {
                cachedApps = new ArrayList<>();
                List<PackageInfo> packageInfos = pm.getInstalledPackages(PackageManager.GET_PERMISSIONS);
                for (PackageInfo p : packageInfos) {
                    if (p.requestedPermissions != null) {
                        List<String> requestPermissions = Arrays.asList(p.requestedPermissions);
                        if (requestPermissions.contains(Manifest.permission.INTERNET)) {
                            ProxiedApp app = new ProxiedApp(pm.getApplicationLabel(p.applicationInfo).toString(), p.packageName, p.applicationInfo.loadIcon(pm));
                            cachedApps.add(app);
                        }
                    }
                }
            }
        }
        return cachedApps;
    }

    public class ProxiedApp {
        public String name;
        public String packageName;
        public Drawable icon;

        public ProxiedApp(String name, String packageName, Drawable icon) {
            this.name = name;
            this.packageName = packageName;
            this.icon = icon;
        }
    }

    private class ListEntry {
        public Switch mSwitch;
        public TextView text;
        public ImageView icon;

        public ListEntry(Switch sw, TextView text, ImageView icon) {
            this.mSwitch = sw;
            this.text = text;
            this.icon = icon;
        }
    }

    private class AppViewHolder extends RecyclerView.ViewHolder implements View.OnClickListener {

        private ImageView icon;
        private Switch check;
        private ProxiedApp item;

        public AppViewHolder(View view) {
            super(view);
            icon = itemView.findViewById(R.id.itemicon);
            check = itemView.findViewById(R.id.itemcheck);

            itemView.setOnClickListener(this);
        }

        private boolean proxied() {
            return proxiedApps.contains(item.packageName);
        }

        public void bind(ProxiedApp app) {
            this.item = app;
            icon.setImageDrawable(app.icon);
            check.setText(app.name);
            check.setChecked(proxied());
        }

        @Override
        public void onClick(View v) {
            if (proxied()) {
                proxiedApps.remove(item.packageName);
                check.setChecked(false);
            } else {
                proxiedApps.add(item.packageName);
                check.setChecked(true);
            }
            if (!appsLoading.get()) {
                profile.individual = Utils.makeString(proxiedApps, "\n");
                ShadowsocksApplication.app.profileManager.updateProfile(profile);
            }
        }
    }

    private class AppsAdapter extends RecyclerView.Adapter<AppViewHolder> {

        private List<ProxiedApp> apps;

        public AppsAdapter() {
            List<ProxiedApp> apps = getApps(getPackageManager());
            Collections.sort(apps, new Comparator<ProxiedApp>() {
                @Override
                public int compare(ProxiedApp a, ProxiedApp b) {
                    boolean aProxied = proxiedApps.contains(a.packageName);
                    if (aProxied ^ proxiedApps.contains(b.packageName)) {
                        return aProxied ? -1 : 1;
                    } else {
                        boolean result = a.name.compareToIgnoreCase(b.name) < 0;
                        return result ? 1 : -1;
                    }
                }
            });
            this.apps = apps;

        }

        @Override
        public int getItemCount() {
            return apps == null ? 0 : apps.size();
        }

        @Override
        public void onBindViewHolder(AppViewHolder vh, int i) {
            vh.bind(apps.get(i));
        }

        @Override
        public AppViewHolder onCreateViewHolder(ViewGroup vg, int viewType) {
            View view = LayoutInflater.from(vg.getContext()).inflate(R.layout.layout_apps_item, vg, false);
            return new AppViewHolder(view);
        }
    }
}

File: app/src/main/java/com/bige0/shadowsocksr/AppManager.kt
package com.bige0.shadowsocksr

import android.*
import android.app.*
import android.content.*
import android.content.pm.*
import android.graphics.drawable.*
import android.os.*
import android.view.*
import android.widget.*
import androidx.appcompat.app.*
import androidx.appcompat.widget.Toolbar
import androidx.recyclerview.widget.*
import com.bige0.shadowsocksr.database.*
import com.bige0.shadowsocksr.utils.*
import java.util.*
import java.util.concurrent.atomic.*

class AppManager : AppCompatActivity(), Toolbar.OnMenuItemClickListener
{
	private var receiverRegistered: Boolean = false
	private var cachedApps = mutableListOf<ProxiedApp>()

	private var proxiedApps = HashSet<String>()
	private lateinit var toolbar: Toolbar
	private lateinit var bypassSwitch: Switch
	private lateinit var appListView: RecyclerView
	private lateinit var loadingView: View
	private var appsLoading: AtomicBoolean = AtomicBoolean()
	private var handler: Handler? = null
	private val profile: Profile? = ShadowsocksApplication.app.currentProfile()

	private fun initProxiedApps(str: String? = profile?.individual)
	{
		if (str.isNullOrEmpty())
		{
			return
		}
		val split = str.lines()
		proxiedApps = HashSet(split)
	}

	override fun onDestroy()
	{
		super.onDestroy()

		if (handler != null)
		{
			handler!!.removeCallbacksAndMessages(null)
			handler = null
		}
	}

	override fun onMenuItemClick(item: MenuItem): Boolean
	{
		val clipboard = getSystemService(Context.CLIPBOARD_SERVICE) as ClipboardManager?
		val proxiedAppString: String
		when (item.itemId)
		{
			R.id.action_apply_all ->
			{
				val profiles = ShadowsocksApplication.app.profileManager.allProfiles
				if (profiles.isNotEmpty() && profile != null)
				{
					proxiedAppString = profile.individual
					val inProxiedApp = profile.proxyApps
					val inBypass = profile.bypass
					for (p in profiles)
					{
						p.individual = proxiedAppString
						p.bypass = inBypass
						p.proxyApps = inProxiedApp
						ShadowsocksApplication.app.profileManager.updateProfile(p)
					}
					ToastUtils.showShort(R.string.action_apply_all)
				}
				else
				{
					ToastUtils.showShort(R.string.action_export_err)
				}
				return true
			}
			R.id.action_export ->
			{
				if (profile == null || clipboard == null)
				{
					ToastUtils.showShort(R.string.action_export_err)
					return false
				}
				val bypass = profile.bypass
				proxiedAppString = profile.individual
				val clip = ClipData.newPlainText(Constants.Key.individual, "$bypass\n$proxiedAppString")
				clipboard.setPrimaryClip(clip)
				ToastUtils.showShort(R.string.action_export_msg)
				return true
			}
			R.id.action_import ->
			{
				if (clipboard != null && clipboard.hasPrimaryClip() && profile != null)
				{
					val proxiedAppSequence = clipboard.primaryClip!!.getItemAt(0)
						.text
					if (proxiedAppSequence != null)
					{
						proxiedAppString = proxiedAppSequence.toString()
						if (proxiedAppString.isNotEmpty())
						{
							//TODO
							val i = proxiedAppString.indexOf('\n')
							try
							{
								val enabled: String
								val apps: String
								if (i < 0)
								{
									enabled = proxiedAppString
									apps = ""
								}
								else
								{
									enabled = proxiedAppString.substring(0, i)
									apps = proxiedAppString.substring(i + 1)
								}

								bypassSwitch.isChecked = enabled.toBoolean()
								profile.individual = apps
								ShadowsocksApplication.app.profileManager.updateProfile(profile)
								ToastUtils.showShort(R.string.action_import_msg)
								appListView.visibility = View.GONE
								loadingView.visibility = View.VISIBLE
								initProxiedApps(apps)
								reloadApps()
								return true
							}
							catch (e: Exception)
							{
								//ignore
							}
						}
					}
				}
				ToastUtils.showShort(R.string.action_import_err)
				return false
			}
			else ->
			{
			}
		}
		return false
	}

	override fun onCreate(savedInstanceState: Bundle?)
	{
		super.onCreate(savedInstanceState)
		if (profile == null)
		{
			finish()
			return
		}

		handler = Handler()

		setContentView(R.layout.layout_apps)
		toolbar = findViewById(R.id.toolbar)
		toolbar.setTitle(R.string.proxied_apps)
		toolbar.setNavigationIcon(R.drawable.abc_ic_ab_back_material)
		toolbar.setNavigationOnClickListener {
			val intent = parentActivityIntent
			if (shouldUpRecreateTask(intent) || isTaskRoot)
			{
				TaskStackBuilder.create(this@AppManager)
					.addNextIntentWithParentStack(intent)
					.startActivities()
			}
			else
			{
				finish()
			}
		}
		toolbar.inflateMenu(R.menu.app_manager_menu)
		toolbar.setOnMenuItemClickListener(this)

		if (!profile.proxyApps)
		{
			profile.proxyApps = true
			ShadowsocksApplication.app.profileManager.updateProfile(profile)
		}

		(findViewById<View>(R.id.onSwitch) as Switch).setOnCheckedChangeListener { _, checked ->
			profile.proxyApps = checked
			ShadowsocksApplication.app.profileManager.updateProfile(profile)
			finish()
		}

		bypassSwitch = findViewById(R.id.bypassSwitch)
		bypassSwitch.isChecked = profile.bypass
		bypassSwitch.setOnCheckedChangeListener { _, isChecked ->
			profile.bypass = isChecked
			ShadowsocksApplication.app.profileManager.updateProfile(profile)
		}
		initProxiedApps()
		loadingView = findViewById(R.id.loading)
		appListView = findViewById(R.id.applistview)
		appListView.layoutManager = LinearLayoutManager(this)
		appListView.itemAnimator = DefaultItemAnimator()

		Thread { loadApps() }.start()
	}


	private fun reloadApps()
	{
		if (!appsLoading.compareAndSet(true, false))
		{
			loadApps()
		}
	}

	private fun loadApps()
	{
		if (!appsLoading.compareAndSet(false, true))
		{
			return
		}
		var tempAdapter: AppsAdapter
		do
		{
			appsLoading.set(true)
			tempAdapter = AppsAdapter()
		} while (!appsLoading.compareAndSet(true, false))

		handler?.post {
			appListView.adapter = tempAdapter
			Utils.crossFade(this@AppManager, loadingView, appListView)
		}
	}

	override fun onKeyUp(keyCode: Int, event: KeyEvent): Boolean
	{
		return if (keyCode == KeyEvent.KEYCODE_MENU)
		{
			if (toolbar.isOverflowMenuShowing)
			{
				toolbar.hideOverflowMenu()
			}
			else
			{
				toolbar.showOverflowMenu()
			}
		}
		else super.onKeyUp(keyCode, event)
	}

	private fun getApps(pm: PackageManager): MutableList<ProxiedApp>
	{
		if (!receiverRegistered)
		{
			val filter = IntentFilter(Intent.ACTION_PACKAGE_ADDED)
			filter.addAction(Intent.ACTION_PACKAGE_REMOVED)
			filter.addDataScheme("package")
			ShadowsocksApplication.app.registerReceiver(object : BroadcastReceiver()
														{
															override fun onReceive(context: Context, intent: Intent)
															{
																if (Intent.ACTION_PACKAGE_REMOVED == intent.action || !intent.getBooleanExtra(Intent.EXTRA_REPLACING, false))
																{
																	synchronized(ProxiedApp::class) {
																		reloadApps()
																	}
																}
															}
														}, filter)
			receiverRegistered = true
		}

		synchronized(cachedApps) {
			if (cachedApps.isEmpty())
			{
				val packagesInfo = pm.getInstalledPackages(PackageManager.GET_PERMISSIONS)
				for (p in packagesInfo)
				{
					if (p.requestedPermissions != null)
					{
						val requestPermissions = listOf(*p.requestedPermissions)
						if (requestPermissions.contains(Manifest.permission.INTERNET))
						{
							val app = ProxiedApp(pm.getApplicationLabel(p.applicationInfo).toString(), p.packageName, p.applicationInfo.loadIcon(pm))
							cachedApps.add(app)
						}
					}
				}
			}
		}
		return cachedApps
	}

	inner class ProxiedApp(var name: String, var packageName: String, var icon: Drawable)

	private inner class AppViewHolder(view: View) : RecyclerView.ViewHolder(view), View.OnClickListener
	{
		private val icon: ImageView = itemView.findViewById(R.id.itemicon)
		private val check: Switch = itemView.findViewById(R.id.itemcheck)
		private lateinit var item: ProxiedApp

		init
		{
			itemView.setOnClickListener(this)
		}

		private fun proxied(): Boolean
		{
			return proxiedApps.contains(item.packageName)
		}

		fun bind(app: ProxiedApp)
		{
			item = app
			icon.setImageDrawable(app.icon)
			check.text = app.name
			check.isChecked = proxied()
		}

		override fun onClick(v: View)
		{
			if (proxied())
			{
				proxiedApps.remove(item.packageName)
				check.isChecked = false
			}
			else
			{
				proxiedApps.add(item.packageName)
				check.isChecked = true
			}
			if (!appsLoading.get() && profile != null)
			{
				profile.individual = Utils.makeString(proxiedApps, "\n")
				ShadowsocksApplication.app.profileManager.updateProfile(profile)
			}
		}
	}

	private inner class AppsAdapter : RecyclerView.Adapter<AppViewHolder>()
	{
		private val apps: List<ProxiedApp>

		init
		{
			val apps = getApps(packageManager)
			apps.sortWith(Comparator { a, b ->
				val aProxied = proxiedApps.contains(a.packageName)
				if (aProxied xor proxiedApps.contains(b.packageName))
				{
					if (aProxied) -1 else 1
				}
				else
				{
					val result = a.name.compareTo(b.name, ignoreCase = true) < 0
					if (result) 1 else -1
				}
			})
			this.apps = apps
		}

		override fun getItemCount(): Int
		{
			return apps.size
		}

		override fun onBindViewHolder(vh: AppViewHolder, i: Int)
		{
			vh.bind(apps[i])
		}

		override fun onCreateViewHolder(vg: ViewGroup, viewType: Int): AppViewHolder
		{
			val view = LayoutInflater.from(vg.context)
				.inflate(R.layout.layout_apps_item, vg, false)
			return AppViewHolder(view)
		}
	}
}
